TechSmith/ThirdParty-Packages-vcpkg
# Add ffmpeg build for cloud team, with mjpeg & png decoding support

# Description
The cloud team needs a build of ffmpeg with mjpeg encoding support.  This adds support for this package, and will ONLY publish the files they need.

This includes adding support to optionally publish the "tools" dir, and support to optionally NOT publish other dirs and files.

# Testing

## Test: Extract frames as png files from TSC2-encoded avi
1. Download the windows release and unzip it: https://github.com/TechSmith/ThirdParty-Packages-vcpkg/releases/download/ffmpeg-cloud-6.1.1--20241002-1135/ffmpeg-cloud-windows-release-bin.tar.gz
2. Download this file to the unzipped dir from that release: [RelayProfileTest.avi](https://github.com/TechSmith/Luma.NET-SampleCode/blob/8f981cecf2820e3fabc86d5ded5264f3b00ee998/Luma.NetCore.ConsoleApp/TestMedia/RelayProfileTest.avi)
3. In a powershell command prompt in that unzipped dir from that release, run this command: `.\ffmpeg.exe -i .\RelayProfileTest.avi -vf fps=1 -f image2 RelayProfileTestFrame%06d.png`
4. Verify it outputs 3 png files, which look like this:

<img width="50%" src="https://github.com/user-attachments/assets/6504bfac-a7ee-4f81-952d-6598d5749460"/>

## Test: Extract frames as PNGs from an h.264-encoded mp4
1. Download the release and unzip it: https://github.com/TechSmith/ThirdParty-Packages-vcpkg/releases/download/ffmpeg-cloud-6.1.1--20241002-1135/ffmpeg-cloud-windows-release-bin.tar.gz
2. Download this file to the unzipped dir from that release: [HarryPotter.mp4](https://github.com/TechSmith/TSCTestMediaFiles/blob/main/HarryPotter.mp4)
3. In a powershell command prompt in that unzipped dir from that release, run this command: `.\ffmpeg.exe -i .\HarryPotter.mp4 -vf fps=30 HarryPotterFrame%06d.png`
4. Verify that 600 png files are being created, one for every frame of the video

Example (HarryPotterFrame000022.png)
<img width="50%" src="https://github.com/user-attachments/assets/1dcb1e2a-86c0-434a-bf86-1794dc7f596a" />

## Test: Output 1 frame every 2 seconds as PNG
1. Download the release and unzip it: https://github.com/TechSmith/ThirdParty-Packages-vcpkg/releases/download/ffmpeg-cloud-6.1.1--20241002-1135/ffmpeg-cloud-windows-release-bin.tar.gz
2. Download this file to the unzipped dir from that release: [HarryPotter.mp4](https://github.com/TechSmith/TSCTestMediaFiles/blob/main/HarryPotter.mp4)
3. In a powershell command prompt in that unzipped dir from that release, run these commands: 
  - `rm -Recurse -Force *.png`
  - `ffmpeg.exe -i "HarryPotter.mp4" -vf fps=0.5 "HarryPotter%06d.png"`
4. Verify that 10 png files are being created, one for every 2 seconds of the video

## Test: Generate jpg thumbnail at a certain time
1. Download the release and unzip it: https://github.com/TechSmith/ThirdParty-Packages-vcpkg/releases/download/ffmpeg-cloud-6.1.1--20241002-1135/ffmpeg-cloud-windows-release-bin.tar.gz
2. Download this file to the unzipped dir from that release: [HarryPotter.mp4](https://github.com/TechSmith/TSCTestMediaFiles/blob/main/HarryPotter.mp4)
3. In a powershell command prompt in that unzipped dir from that release, run these commands: 
  - `rm -Recurse -Force *.png`
  - `ffmpeg.exe -ss 3 -i "HarryPotter.mp4" -vframes 1 "HarryPotterThumb.jpg"`
4. Verify that one png file is created, HarryPotterThumb, that looks like this:

<img width="50%" src="https://github.com/user-attachments/assets/63c926a1-768e-49c5-a9bb-9451e08ed1f2" />

## Test: Transcode video to different resolution
1. Download the release and unzip it: https://github.com/TechSmith/ThirdParty-Packages-vcpkg/releases/download/ffmpeg-cloud-6.1.1--20241002-1135/ffmpeg-cloud-windows-release-bin.tar.gz
2. Download this file to the unzipped dir from that release: [3840x2160_H264_MP4.mp4](https://github.com/TechSmith/TSCTestMediaFiles/blob/main/3840x2160_H264_MP4.mp4)
3. In a powershell command prompt in that unzipped dir from that release, run these commands: 
  - `rm -Recurse -Force *.png`
  - `ffmpeg.exe -i "3840x2160_H264_MP4.mp4" -movflags faststart -r 30 -force_key_frames 'expr:gte(t,n_forced*2)' -video_track_timescale 30000 -map_metadata 0 -vf scale=1920:1080 "3840x2160_H264_MP4-Transcoded.mp4" -y`
4. Run this command: `ffprobe.exe .\3840x2160_H264_MP4-Transcoded.mp4`
5. Verify you see that it is transcoded to h.264 at 1920x1080:

```
PS C:\Users\m.malburg\Downloads\ffmpeg-cloud-windows-release-bin (1)> ffprobe .\3840x2160_H264_MP4-Transcoded.mp4
ffprobe version 6.1.1-essentials_build-www.gyan.dev Copyright (c) 2007-2023 the FFmpeg developers
  built with gcc 12.2.0 (Rev10, Built by MSYS2 project)
  configuration: --enable-gpl --enable-version3 --enable-static --pkg-config=pkgconf --disable-w32threads --disable-autodetect --enable-fontconfig --enable-iconv --enable-gnutls --enable-libxml2 --enable-gmp --enable-bzlib --enable-lzma --enable-zlib --enable-libsrt --enable-libssh --enable-libzmq --enable-avisynth --enable-sdl2 --enable-libwebp --enable-libx264 --enable-libx265 --enable-libxvid --enable-libaom --enable-libopenjpeg --enable-libvpx --enable-mediafoundation --enable-libass --enable-libfreetype --enable-libfribidi --enable-libharfbuzz --enable-libvidstab --enable-libvmaf --enable-libzimg --enable-amf --enable-cuda-llvm --enable-cuvid --enable-ffnvcodec --enable-nvdec --enable-nvenc --enable-dxva2 --enable-d3d11va --enable-libvpl --enable-libgme --enable-libopenmpt --enable-libopencore-amrwb --enable-libmp3lame --enable-libtheora --enable-libvo-amrwbenc --enable-libgsm --enable-libopencore-amrnb --enable-libopus --enable-libspeex --enable-libvorbis --enable-librubberband
  libavutil      58. 29.100 / 58. 29.100
  libavcodec     60. 31.102 / 60. 31.102
  libavformat    60. 16.100 / 60. 16.100
  libavdevice    60.  3.100 / 60.  3.100
  libavfilter     9. 12.100 /  9. 12.100
  libswscale      7.  5.100 /  7.  5.100
  libswresample   4. 12.100 /  4. 12.100
  libpostproc    57.  3.100 / 57.  3.100
Input #0, mov,mp4,m4a,3gp,3g2,mj2, from '.\3840x2160_H264_MP4-Transcoded.mp4':
  Metadata:
    major_brand     : isom
    minor_version   : 512
    compatible_brands: isomiso2avc1mp41
    creation_time   : 2016-12-12T20:51:49.000000Z
    encoder         : Lavf60.16.100
  Duration: 00:00:07.01, start: 0.000000, bitrate: 1520 kb/s
  Stream #0:0[0x1](und): Video: h264 (High) (avc1 / 0x31637661), yuv420p(progressive), 1920x1080 [SAR 1:1 DAR 16:9], 1377 kb/s, 30 fps, 30 tbr, 30k tbn (default)
    Metadata:
      creation_time   : 2016-12-12T20:51:49.000000Z
      handler_name    : Core Media Video
      vendor_id       : [0][0][0][0]
      encoder         : Lavc60.31.102 libx264
    Side data:
      ICC Profile
  Stream #0:1[0x2](und): Audio: aac (LC) (mp4a / 0x6134706D), 44100 Hz, stereo, fltp, 131 kb/s (default)
    Metadata:
      creation_time   : 2016-12-12T20:51:49.000000Z
      handler_name    : Core Media Audio
      vendor_id       : [0][0][0][0]
```
